salt/modules/opkg.py

```diff
@@ -370,15 +374,175 @@ def _parse_reported_packages_from_install_output(output):
+
+    Example:
+        operation = 'install'
+        output = 'Installing vim (17.0.0) on root'
+    Returns: vim
+    '''
+    package = None
+    if operation == 'download' and output.endswith('.ipk.'):
+        package = output.split('/')[-1].split('_')[0]
+    elif operation in ['install', 'upgrade', 'remove', 'downgrade']:
+        output_split = output.split()
+        # For example: "Installing/Removing/.. package_name (version_number) ..."
+        # We do not want to count "Removing any package that.." messages
+        if len(output_split) >= 3 and output_split[2].startswith('('):
+            package = output_split[1]
+    
+    return package
+
+def _process_with_progress(cmd, jid, total_packages_count):
+    '''
+    Process the opkg command and fire events to the salt-master bus
+    indicating the progress of the operation
+    '''
+    out = {}
+    stdout = ''
+    minion_id = str(salt.config.get_id(__opts__)[0])
+    notify_progress_period = __opts__.get('pkg_progress_period', 60)
+    last_notify_timestamp = 0
+    processed_count = 0
+    last_processing_package = None
+    last_or_first_package = True
+    operation = None
+    event = salt.utils.event.get_event(
+            'minion', opts=__opts__, listen=False
+            )
+    tag = 'salt/minion/pkg_progress/{0}'.format(minion_id)
+    proc = subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            bufsize=1,
+            universal_newlines=True
+            )
+    # We don't want the standard output to be blocking to avoid missing updates
+    # based on specified progress period
+    flags = fcntl(proc.stdout, F_GETFL)
+    fcntl(proc.stdout.fileno(), F_SETFL, flags | os.O_NONBLOCK)
+    while True:
+        ready_to_read = select.select([proc.stdout.fileno()], [], [])[0]
+        if ready_to_read:
+            output = proc.stdout.readline()
+        else:
+            output = ''
+            time.sleep(1)
+        if output == '' and proc.poll() is not None:
+            break
+        if output:
+            stdout += output
+            if not _is_valid_operation_line(output):
+                continue
+            operation = _get_operation_from_output(output)
+            if operation == 'download':
+                last_processing_package = _get_package_from_output(operation, output)
+            elif operation is not None:
+                last_processing_package = _get_package_from_output(operation, output)
+                if last_processing_package is not None:
+                    processed_count += 1
+                    if processed_count == total_packages_count:
+                        last_or_first_package = True
+        timestamp = time.time()
+        if (timestamp - last_notify_timestamp > notify_progress_period or last_or_first_package) and operation is not None and last_processing_package is not None:
```

Comment:
It is inside the while for me?
Also the check for operation and current_package is needed, because it's not inside the if that "continues", it's this way to not block if the last timestamp was 0.99 (for example for a pkg_notify_period of 1), and the current operation takes 6 seconds - so it sends the notification even though there is no new output